torch/csrc/jit/serialization/callstack_debug_info_serialization.cpp

```diff
@@ -0,0 +1,189 @@
+#include <torch/csrc/jit/api/compilation_unit.h>
+#include <torch/csrc/jit/serialization/callstack_debug_info_serialization.h>
+#include <torch/csrc/jit/serialization/pickle.h>
+
+namespace torch {
+namespace jit {
+
+c10::IValue InlinedCallStackSerializer::serialize(
+    const InlinedCallStackPtr& cs_ptr,
+    const SourceRangeTagMap& source_range_tags) {
+  if (!cs_ptr) {
+    return c10::IValue();
+  }
+  auto cs_it = serialized_inlined_callstack_.find(cs_ptr);
+  if (cs_it != serialized_inlined_callstack_.end()) {
+    return cs_it->second;
+  }
+  // Inlined callstack pointer is serialized as tuple of 3 elements
+  // {IValue(module_instance_info), source_range_tag, IValue(InlinedCallStack)}
+  std::vector<c10::IValue> elements;
+  elements.reserve(3);
+  elements.emplace_back(
+      serialize_module_instance_info(cs_ptr->module_instance()));
+  int64_t source_range_tag{-1};
+  const SourceRange& sr = cs_ptr->source_range().findSourceRangeThatGenerated()
+      ? cs_ptr->source_range().findSourceRangeThatGenerated().value()
+      : cs_ptr->source_range();
+  auto sr_it = source_range_tags.find(sr);
+  if (sr_it != source_range_tags.end()) {
+    source_range_tag = sr_it->second;
+  }
+  elements.emplace_back(source_range_tag);
+  if (cs_ptr->callee()) {
+    elements.emplace_back(
+        serialize(cs_ptr->callee().value(), source_range_tags));
+  } else {
+    elements.emplace_back(c10::IValue());
+  }
+  c10::IValue serialized_cs = c10::ivalue::Tuple::create(elements);
+  serialized_inlined_callstack_[cs_ptr] = serialized_cs;
+  return serialized_cs;
+}
+
+c10::IValue InlinedCallStackSerializer::serialize_module_instance_info(
+    const c10::optional<ModuleInstanceInfo>& m) {
+  if (!m) {
+    return c10::IValue();
+  }
+  const auto& m_val = m.value();
+  std::vector<c10::IValue> elements;
+  // Module instance info is serialized as
+  // {type name, instance name}
+  // Note that the same module instance will appear in many callstacks.
+  // So it makes sense to cache these and use cached value.
+  // TODO: Add caching for moudle instance info serialization.
+  if (m_val.class_type()) {
+    elements = {
+        m_val.class_type()->name()->qualifiedName(), m_val.instance_name()};
+  } else {
+    elements = {"", m_val.instance_name()};
+  }
+  return c10::ivalue::Tuple::create(std::move(elements));
+}
+
+std::vector<char> CallStackDebugInfoPickler::pickle(
+    const std::unordered_map<int64_t, DebugInfoPair>& callstack_ptrs,
+    const SourceRangeTagMap& source_range_tags) {
+  std::vector<c10::IValue> ivalues;
+  for (const auto& it : callstack_ptrs) {
+    int64_t debug_handle = it.first;
+    std::vector<c10::IValue> elements;
+    /*
+     * Debug handles and debug info (source range + inlinded callstack)
+     * are serialized as a tuple of 3 elements
+     * {debug_handle, source_range_tag, serialized_callstack}
```

Comment:
curious - why do we serialize the 'source_range_tag' here again? does it uniquely identify a callstack?

Reply:
source_range_tag does not uniquely identify a call stack. Debug handle does that. Source_range_tag just tell you about source location. So like in callstack each entry will tell you where in the source code it is. That is why each callstack entry needs source_range_tag.

Reply:
So this source_range_tag (DebugInfoPair.first) is the very first entry that's not included in the callstack list (DebugInfoPair.second)? Seems annoying that we have to handle these two types of callstacks (top v.s. non-top) separately at multiple places :)

Reply:
Yes. I added a comment to the effect here: https://github.com/pytorch/pytorch/pull/55062/files/27d9a6ed14ee39a348b2239678f2f487e4f508e0#diff-1b393ba50e3566ffbda0d8a593ba0658567085279c6fc74ba4328688af877f4a. I tried hvaing just the inlined callstack including top node but soon realized to make it work I will have to change a lot of places in JIT.